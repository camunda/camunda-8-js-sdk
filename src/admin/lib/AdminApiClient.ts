import d from 'debug'
import got from 'got'

import {
	CamundaEnvironmentConfigurator,
	CamundaPlatform8Configuration,
	DeepPartial,
	GetCustomCertificateBuffer,
	GotRetryConfig,
	RequireConfiguration,
	constructOAuthProvider,
	createUserAgentString,
	gotBeforeErrorHook,
	gotErrorHandler,
	makeBeforeRetryHandlerFor401TokenRetry,
} from '../../lib'
import { IOAuthProvider } from '../../oauth'

import * as Dto from './AdminDto'

const debug = d('camunda:adminconsole')

/**
 * This class provides methods to interact with the Camunda Admin API.
 * @throws {RESTError} An error that may occur during API operations.
 */
export class AdminApiClient {
	private userAgentString: string
	private oAuthProvider: IOAuthProvider
	private rest: Promise<typeof got>

	constructor(options?: {
		config?: DeepPartial<CamundaPlatform8Configuration>
		oAuthProvider?: IOAuthProvider
	}) {
		const config = CamundaEnvironmentConfigurator.mergeConfigWithEnvironment(
			options?.config ?? {}
		)
		const prefixUrl = RequireConfiguration(
			config.CAMUNDA_CONSOLE_BASE_URL,
			'CAMUNDA_CONSOLE_BASE_URL'
		)

		this.oAuthProvider =
			options?.oAuthProvider ?? constructOAuthProvider(config)

		this.userAgentString = createUserAgentString(config)
		this.rest = GetCustomCertificateBuffer(config).then(
			(certificateAuthority) =>
				got.extend({
					prefixUrl,
					retry: GotRetryConfig,
					https: {
						certificateAuthority,
					},
					handlers: [gotErrorHandler],
					hooks: {
						beforeRetry: [
							makeBeforeRetryHandlerFor401TokenRetry(
								this.getHeaders.bind(this)
							),
						],
						beforeError: [gotBeforeErrorHook],
						beforeRequest: config.middleware ?? [],
					},
				})
		)
		debug('prefixUrl', `${prefixUrl}`)
	}

	private async getHeaders() {
		const authorization = await this.oAuthProvider.getToken('CONSOLE')
		const headers = {
			'content-type': 'application/json',
			...authorization,
			'user-agent': this.userAgentString,
			accept: '*/*',
		}
		return headers
	}

	/**
	 *
	 * @description Get an array of the current API clients for this cluster. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/GetClients) for more details.
	 * @throws {RESTError}
	 * @param clusterUuid - The cluster UUID
	 *
	 */
	async getClients(clusterUuid: string): Promise<Dto.ClusterClient[]> {
		const headers = await this.getHeaders()
		const rest = await this.rest
		return rest(`clusters/${clusterUuid}/clients`, {
			headers,
		}).json()
	}

	/**
	 * @description Create a new API client for a cluster. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/CreateClient) for more details.
	 * @throws {RESTError}
	 */
	async createClient(req: {
		clusterUuid: string
		clientName: string
		permissions: string[]
	}): Promise<Dto.CreatedClusterClient> {
		const headers = await this.getHeaders()
		const rest = await this.rest
		return rest
			.post(`clusters/${req.clusterUuid}/clients`, {
				body: JSON.stringify({
					clientName: req.clientName,
					permissions: req.permissions,
				}),
				headers,
			})
			.json()
	}

	/**
	 * @description Get the details of an API client. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/GetClient) for more details.
	 * @param clusterUuid
	 * @param clientId
	 * @throws {RESTError}
	 * @returns
	 */
	async getClient(
		clusterUuid: string,
		clientId: string
	): Promise<Dto.ClusterClientConnectionDetails> {
		const headers = await this.getHeaders()
		const rest = await this.rest
		return rest(`clusters/${clusterUuid}/clients/${clientId}`, {
			headers,
		}).json()
	}

	/**
	 * @description See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/DeleteClient) for more details.
	 * @param clusterUuid
	 * @param clientId
	 * @throws {RESTError}
	 */
	async deleteClient(clusterUuid: string, clientId: string): Promise<null> {
		const headers = await this.getHeaders()
		const rest = await this.rest
		return rest
			.delete(`clusters/${clusterUuid}/clients/${clientId}`, {
				headers,
			})
			.json()
	}

	/**
	 *
	 * @description Return an array of clusters. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/GetClusters) for more details.
	 * @throws {RESTError}
	 */
	async getClusters(): Promise<Dto.Cluster[]> {
		const headers = await this.getHeaders()
		const rest = await this.rest
		return rest('clusters', {
			headers,
		}).json()
	}

	/**
	 *
	 * @description Create a new cluster. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/CreateCluster) for more details.
	 * @throws {RESTError}
	 */
	async createCluster(
		createClusterRequest: Dto.CreateClusterBody
	): Promise<{ clusterId: string }> {
		const headers = await this.getHeaders()
		const req = {
			body: JSON.stringify(createClusterRequest),
			headers,
		}
		const rest = await this.rest
		return rest.post('clusters', req).json()
	}

	/**
	 *
	 * @description Retrieve the metadata for a cluster. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/GetCluster) for more details.
	 * @throws {RESTError}
	 *
	 */
	async getCluster(clusterUuid: string): Promise<Dto.Cluster> {
		const headers = await this.getHeaders()
		const rest = await this.rest
		return rest(`clusters/${clusterUuid}`, {
			headers,
		}).json()
	}

	/**
	 *
	 * @description Delete a cluster. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/DeleteCluster) for more details.
	 * @throws {RESTError}
	 *
	 */
	async deleteCluster(clusterUuid: string): Promise<null> {
		const headers = await this.getHeaders()
		const rest = await this.rest
		return rest
			.delete(`clusters/${clusterUuid}`, {
				headers,
			})
			.json()
	}

	/**
	 *
	 * @description Retrieve the available parameters for cluster creation. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/GetParameters) for more details.
	 * @throws {RESTError}
	 */
	async getParameters(): Promise<Dto.Parameters> {
		const headers = await this.getHeaders()
		const rest = await this.rest
		return rest('clusters/parameters', {
			headers,
		}).json()
	}

	/**
	 *
	 * @description Retrieve the connector secrets. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/GetSecrets) for more details.
	 * @throws {RESTError}
	 */
	async getSecrets(clusterUuid: string): Promise<{ [key: string]: string }> {
		const headers = await this.getHeaders()
		const rest = await this.rest
		return rest(`clusters/${clusterUuid}/secrets`, {
			headers,
		}).json()
	}

	/**
	 *
	 * @description Create a new connector secret. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/CreateSecret) for more details.
	 * @throws {RESTError}
	 */
	async createSecret({
		clusterUuid,
		secretName,
		secretValue,
	}: {
		clusterUuid: string
		secretName: string
		secretValue: string
	}): Promise<null> {
		const headers = await this.getHeaders()
		const req = {
			body: JSON.stringify({ secretName, secretValue }),
			headers,
		}
		const rest = await this.rest
		return rest.post(`clusters/${clusterUuid}/secrets`, req).json()
	}

	/**
	 *
	 * @description Delete a connector secret. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/DeleteSecret) for more details.
	 * @throws {RESTError}
	 */
	async deleteSecret(clusterUuid: string, secretName: string): Promise<null> {
		const headers = await this.getHeaders()
		const rest = await this.rest
		return rest
			.delete(`clusters/${clusterUuid}/secrets/${secretName}`, {
				headers,
			})
			.json()
	}

	/**
	 *
	 * @description Add one or more IPs to the whitelist for the cluster. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/UpdateIpWhitelist) for more details.
	 * @throws {RESTError}
	 * @param ipwhitelist
	 * @returns
	 */
	async whitelistIPs(
		clusterUuid: string,
		ipwhitelist: [
			{
				description: string
				ip: string
			},
		]
	): Promise<null> {
		const headers = await this.getHeaders()
		const rest = await this.rest
		return rest
			.put(`clusters/${clusterUuid}/ipwhitelist`, {
				body: JSON.stringify({
					ipwhitelist,
				}),
				headers,
			})
			.json()
	}

	/**
	 *
	 * @description Retrieve a list of members and pending invites for your organisation. See the [API Documentation]() for more details.
	 * @throws {RESTError}
	 */
	async getUsers(): Promise<Dto.Member[]> {
		const headers = await this.getHeaders()
		const rest = await this.rest
		return rest
			.get('members', {
				headers,
			})
			.json()
	}

	/**
	 *
	 * @description Add a member. See the [API Documentation]() for more details.
	 * @throws {RESTError}
	 *
	 */
	async createMember(
		email: string,
		orgRoles: Dto.OrganizationRole[]
	): Promise<null> {
		const headers = await this.getHeaders()
		const rest = await this.rest
		return rest
			.post(`members/${email}`, {
				headers,
				body: JSON.stringify({ orgRoles }),
			})
			.json()
	}

	/**
	 *
	 * @description Delete a member from your organization. See the [API Documentation]() for more details.
	 * @throws {RESTError}
	 *
	 */
	async deleteMember(email: string): Promise<null> {
		const headers = await this.getHeaders()
		const rest = await this.rest
		return rest
			.delete(`members/${email}`, {
				headers,
			})
			.json()
	}
}
