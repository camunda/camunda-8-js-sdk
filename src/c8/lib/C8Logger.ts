import winston from 'winston' // Import Winston

import {
	Camunda8ClientConfiguration,
	CamundaEnvironmentConfigurator,
} from '../../lib'

export type Logger = {
	/* eslint-disable @typescript-eslint/no-explicit-any */
	info: (message: string | undefined, ...meta: any[]) => void
	warn: (message: string | undefined, ...meta: any[]) => void
	error: (message: string | undefined, ...meta: any[]) => void
	debug: (message: string | undefined, ...meta: any[]) => void
	trace: (message: string | undefined, ...meta: any[]) => void
	/* eslint-enable @typescript-eslint/no-explicit-any */
}

let defaultLogger: Logger
let cachedLogger: Logger | undefined

export function getLogger(config?: Camunda8ClientConfiguration) {
	const configuration =
		CamundaEnvironmentConfigurator.mergeConfigWithEnvironment(config ?? {})
	// We assume that the SDK user uses a single winston instance for 100% of logging, or no logger at all (in which case we create our own)
	if (config?.logger && cachedLogger !== config.logger) {
		cachedLogger = config.logger
		config.logger.debug(`Using supplied logger`)
	}
	if (!defaultLogger) {
		// Define the default logger
		const logger: winston.Logger & {
			// eslint-disable-next-line @typescript-eslint/no-explicit-any
			trace: (message: string | undefined, ...meta: any[]) => void
		} = winston.createLogger({
			level: configuration.CAMUNDA_LOG_LEVEL,
			format: winston.format.combine(
				winston.format.timestamp(),
				winston.format.colorize(),
				winston.format.simple()
			),
			transports: [new winston.transports.Console()],
		}) as any // eslint-disable-line @typescript-eslint/no-explicit-any

		logger.trace = logger.silly
		defaultLogger = logger
	}
	if (!cachedLogger) {
		defaultLogger.debug(`Using default winston logger`)
		cachedLogger = defaultLogger
	}
	return config?.logger ?? defaultLogger
}
