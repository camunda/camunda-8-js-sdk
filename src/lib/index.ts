export * from './ClientConstructor'
export * from './Configuration'
export * from './ConstructOAuthProvider'
export * from './CreateDtoInstance'
export * from './CreateUserAgentString'
export * from './Delay'
export * from './EnvironmentSetup'
export * from './GetCustomCertificateBuffer'
export { packageVersion } from './GetPackageVersion'
export * from './GotErrors'
export * from './GotHooks'
export * from './LosslessJsonParser'
export { RequireConfiguration } from './RequireConfiguration'
export * from './ValueOrDefault'
