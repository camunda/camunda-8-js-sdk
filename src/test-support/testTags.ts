/**
 * Test environment selection utilities (Matrix + Exclude model)
 *
 * BREAKING CHANGE: Previous TagFilter / matrix(filterObjectReturningBoolean) semantics removed.
 * New primary APIs:
 *   - matrix({ include: { versions: ['8.8','8.7'], deployments: ['saas','self-managed'] }, exclude: [{ version: '8.8', deployment: 'self-managed' }] })
 *   - allowAny([{ version: '8.7' }, { version: '8.8', deployment: 'saas' }])
 *   - matrix(predicate) -> returns boolean for use with Vitest test.if(matrix(...))
 *
 * Semantics:
 *   matrix(): Produces the cartesian product of provided include dimension arrays. Any dimensions omitted are treated as wildcards (not expanded).
 *             Exclusion patterns (partial matches) remove matching tuples.
 *   allowAny(): Returns true if the current environment matches ANY provided partial pattern.
 *   matrix(): Just invokes the predicate with the current environment and returns a boolean (thin wrapper for clarity / future debug hooks).
 *
 * Example (run on 8.7 all deployments + 8.8 saas only):
 *   const pred = matrix({
 *     include: { versions: ['8.7','8.8'], deployments: ['saas','self-managed'] },
 *     exclude: [{ version: '8.8', deployment: 'self-managed' }]
 *   })
 *   // In a test file (Vitest):
 *   // test.if(matrix(pred))('my test', () => { ...test logic })
 *
 * Example using allowAny (same intent):
 *   const pred = allowAny([{ version: '8.7' }, { version: '8.8', deployment: 'saas' }])
 *   test.if(matrix(pred))('my test', () => {})
 *
 * Debugging: set TEST_TAG_DEBUG=1 to log evaluation decisions.
 */

export const VersionTags = {
	['8.8']: '8.8',
	['8.7']: '8.7',
} as const
export type VersionTag = (typeof VersionTags)[keyof typeof VersionTags]

const DeploymentModeTags = {
	saas: 'saas',
	selfManaged: 'self-managed',
	unitTest: 'unit-test',
} as const
export type DeploymentModeTag =
	(typeof DeploymentModeTags)[keyof typeof DeploymentModeTags]

const TenancyTags = {
	multiTenant: 'multi-tenant',
	singleTenant: 'single-tenant',
} as const
export type TenancyTag = (typeof TenancyTags)[keyof typeof TenancyTags]

const SecurityTags = {
	secured: 'secured',
	unsecured: 'unsecured',
} as const
export type SecurityTag = (typeof SecurityTags)[keyof typeof SecurityTags]

export const TestTags = {
	...VersionTags,
	...DeploymentModeTags,
	...TenancyTags,
	...SecurityTags,
}

// Filter object interface
export interface TestEnvironment {
	version: VersionTag
	deployment: DeploymentModeTag
	tenancy: TenancyTag
	security: SecurityTag
}

export type EnvPattern = Partial<TestEnvironment>

export interface MatrixInclude {
	versions?: VersionTag[]
	deployments?: DeploymentModeTag[]
	tenancy?: TenancyTag[]
	security?: SecurityTag[]
}

export interface MatrixRules {
	include: MatrixInclude
	exclude?: EnvPattern[]
}

export const currentEnv = (): TestEnvironment => ({
	version: (process.env.TEST_VERSION as VersionTag) || '8.8',
	deployment:
		(process.env.TEST_DEPLOYMENT as DeploymentModeTag) || 'self-managed',
	tenancy: (process.env.TEST_TENANCY as TenancyTag) || 'single-tenant',
	security: (process.env.TEST_SECURITY as SecurityTag) || 'unsecured',
})

function matchesPattern(pattern: EnvPattern, env: TestEnvironment): boolean {
	return Object.entries(pattern).every(([k, v]) => (env as never)[k] === v)
}

// Generate predicate from matrix rules
export function matrix(rules: MatrixRules): boolean {
	const includeKeys = Object.entries(rules.include) as [
		keyof MatrixInclude,
		string[],
	][]
	const expanded: EnvPattern[] = []

	if (includeKeys.length === 0) {
		// No explicit include lists given -> everything allowed unless excluded
		expanded.push({})
	} else {
		// Map plural include keys to exact environment property names
		const includeKeyToEnvKey: Record<
			keyof MatrixInclude,
			keyof TestEnvironment
		> = {
			versions: 'version',
			deployments: 'deployment',
			tenancy: 'tenancy',
			security: 'security',
		}
		const recur = (i: number, acc: EnvPattern) => {
			if (i === includeKeys.length) {
				expanded.push(acc)
				return
			}
			const [includeKey, values] = includeKeys[i]
			const envKey = includeKeyToEnvKey[includeKey]
			for (const val of values) {
				recur(i + 1, { ...acc, [envKey]: val } as EnvPattern)
			}
		}
		recur(0, {})
	}

	const env = currentEnv()

	const debug = process.env.TEST_TAG_DEBUG === '1'
	const allowed = expanded.some((p) => matchesPattern(p, env))
	if (!allowed) {
		if (debug)
			console.log('[tagged][matrix] not in include set', { env, expanded })
		return false
	}
	if (rules.exclude?.some((ex) => matchesPattern(ex, env))) {
		if (debug) console.log('[tagged][matrix] excluded', { env })
		return false
	}
	if (debug) console.log('[tagged][matrix] allowed', { env })
	return true
}

// Whitelist of partial environment tuples (OR logic)
export function allowAny(patterns: EnvPattern[]): boolean {
	const env = currentEnv()
	const debug = process.env.TEST_TAG_DEBUG === '1'
	const hit = patterns.some((p) => matchesPattern(p, env))
	if (debug) console.log('[tagged][allowAny]', { env, hit, patterns })
	return hit
}

// Thin wrapper for clarity / future enhancement
export function tagged(predicate: (env: TestEnvironment) => boolean): boolean {
	return predicate(currentEnv())
}

// Export dimension tag objects for caller convenience
export { DeploymentModeTags as DeploymentTags, SecurityTags, TenancyTags }
