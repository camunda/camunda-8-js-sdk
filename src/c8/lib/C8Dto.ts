import { LosslessNumber } from 'lossless-json'

import { Int64String, LosslessDto } from '../../lib'
import { JSONDoc } from '../../zeebe/types'

export class RestApiJob<
	Variables = LosslessDto,
	CustomHeaders = LosslessDto,
> extends LosslessDto {
	@Int64String
	key!: string
	type!: string
	@Int64String
	processInstanceKey!: string
	processDefinitionId!: string
	processDefinitionVersion!: number
	@Int64String
	processDefinitionKey!: string
	elementId!: string
	@Int64String
	elementInstanceKey!: string
	customHeaders!: CustomHeaders
	worker!: string
	retries!: number
	@Int64String
	deadline!: string
	variables!: Variables
	tenantId!: string
}

/**
 * JSON object with changed task attribute values.
 */
export interface TaskChangeSet {
	/* The due date of the task. Reset by providing an empty String. */
	dueDate?: Date | string
	/* The follow-up date of the task. Reset by providing an empty String. */
	followUpDate?: Date | string
	/* The list of candidate users of the task. Reset by providing an empty list. */
	candidateUsers?: string[]
	/* The list of candidate groups of the task. Reset by providing an empty list. */
	candidateGroups?: string[]
}

/** JSON object with changed job attribute values. */
export interface JobUpdateChangeset {
	/* The new amount of retries for the job; must be a positive number. */
	retries?: number
	/**  The duration of the new timeout in ms, starting from the current moment. */
	timeout?: number
}

export interface NewUserInfo {
	password: string
	id: number
	username: string
	name: string
	email: string
	enabled: boolean
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Ctor<T> = new (obj: any) => T

export class ProcessDeployment extends LosslessDto {
	processDefinitionId!: string
	processDefinitionVersion!: number
	@Int64String
	processDefinitionKey!: string
	resourceName!: string
	tenantId!: string
}

export class DecisionDeployment extends LosslessDto {
	dmnDecisionId!: string
	version!: number
	@Int64String
	decisionKey!: string
	dmnDecisionName!: string
	tenantId!: string
	dmnDecisionRequirementsId!: string
	@Int64String
	dmnDecisionRequirementsKey!: string
}

export class DecisionRequirementsDeployment extends LosslessDto {
	dmnDecisionRequirementsId!: string
	version!: number
	dmnDecisionRequirementsName!: string
	tenantId!: string
	@Int64String
	dmnDecisionRequirementsKey!: string
	resourceName!: string
}
export class FormDeployment {
	formId!: string
	version!: number
	@Int64String
	formKey!: string
	resourceName!: string
	tenantId!: string
}

export class DeployResourceResponseDto extends LosslessDto {
	@Int64String
	deploymentKey!: string
	deployments!: (
		| { processDefinition: ProcessDeployment }
		| { decisionDefinition: DecisionDeployment }
		| { decisionRequirements: DecisionRequirementsDeployment }
		| { form: FormDeployment }
	)[]
	tenantId!: string
}

export class DeployResourceResponse extends DeployResourceResponseDto {
	processes!: ProcessDeployment[]
	decisions!: DecisionDeployment[]
	decisionRequirements!: DecisionRequirementsDeployment[]
	forms!: FormDeployment[]
}

export class CreateProcessInstanceResponse<T = Record<string, never>> {
	/**
	 * The unique key identifying the process definition (e.g. returned from a process
	 * in the DeployResourceResponse message)
	 */
	@Int64String
	readonly processDefinitionKey!: string
	/**
	 * The BPMN process ID of the process definition
	 */
	readonly processDefinitionId!: string
	/**
	 * The version of the process; set to -1 to use the latest version
	 */
	readonly version!: number
	@Int64String
	readonly processInstanceKey!: string
	/**
	 * the tenant identifier of the created process instance
	 */
	readonly tenantId!: string
	/**
	 * If `awaitCompletion` is true, this will be populated with the output variables. Otherwise, it will be an empty object.
	 */
	readonly variables!: T
}

export interface MigrationMappingInstruction {
	/** The element ID to migrate from. */
	sourceElementId: string
	/** The element ID to migrate into. */
	targetElementId: string
}

/** Migrates a process instance to a new process definition.
 * This request can contain multiple mapping instructions to define mapping between the active process instance's elements and target process definition elements.
 */
export interface MigrationRequest {
	processInstanceKey: string
	/** The key of process definition to migrate the process instance to. */
	targetProcessDefinitionKey: string
	mappingInstructions: MigrationMappingInstruction[]
	/** A reference key chosen by the user that will be part of all records resulting from this operation. Must be > 0 if provided. */
	operationReference?: number | LosslessNumber
}

/** The signal was broadcast. */
export class BroadcastSignalResponse extends LosslessDto {
	@Int64String
	/** The unique ID of the signal that was broadcast. */
	signalKey!: string
	/** The tenant ID of the signal that was broadcast. */
	tenantId!: string
}

export interface UpdateElementVariableRequest {
	/**
	 * The key of the element instance to update the variables for.
	 * This can be the process instance key (as obtained during instance creation), or a given element,
	 * such as a service task (see the elementInstanceKey on the job message). */
	elementInstanceKey: string
	variables: JSONDoc | LosslessDto
	/**
	 * Defaults to false.
	 * If set to true, the variables are merged strictly into the local scope (as specified by the elementInstanceKey). Otherwise, the variables are propagated to upper scopes and set at the outermost one.
	 * Letâ€™s consider the following example:
	 * There are two scopes '1' and '2'. Scope '1' is the parent scope of '2'. The effective variables of the scopes are: 1 => { "foo" : 2 } 2 => { "bar" : 1 }
	 * An update request with elementInstanceKey as '2', variables { "foo" : 5 }, and local set to true leaves scope '1' unchanged and adjusts scope '2' to { "bar" : 1, "foo" 5 }.
	 * By default, with local set to false, scope '1' will be { "foo": 5 } and scope '2' will be { "bar" : 1 }.
	 */
	local?: boolean
	/**
	 * A reference key chosen by the user that will be part of all records resulting from this operation.
	 * Must be > 0 if provided.
	 */
	operationReference?: number
}

export class CorrelateMessageResponse extends LosslessDto {
	/** the unique ID of the message that was published */
	@Int64String
	key!: string
	/** the tenantId of the message */
	tenantId!: string
	/** The key of the first process instance the message correlated with */
	@Int64String
	processInstanceKey!: string
}

export class PublishMessageResponse extends LosslessDto {
	/** the unique ID of the message that was published */
	@Int64String
	key!: string
	/** the tenantId of the message */
	tenantId!: string
}

export interface CreateProcessBaseRequest<V extends JSONDoc | LosslessDto> {
	/**
	 * the version of the process; if not specified it will use the latest version
	 */
	version?: number
	/**
	 * JSON document that will instantiate the variables for the root variable scope of the
	 * process instance.
	 */
	variables: V
	/** The tenantId for a multi-tenant enabled cluster. */
	tenantId?: string
	/** a reference key chosen by the user and will be part of all records resulted from this operation */
	operationReference?: number | LosslessNumber
	/**
	 * List of start instructions. If empty (default) the process instance
	 * will start at the start event. If non-empty the process instance will apply start
	 * instructions after it has been created
	 */
	startInstructions?: ProcessInstanceCreationStartInstruction[]
	/**
	 * Wait for the process instance to complete. If the process instance completion does not occur within the requestTimeout, the request will be closed. Defaults to false.
	 */
	// This is commented out, because we used specialised methods for the two cases.
	// awaitCompletion?: boolean
	/**
	 * Timeout (in ms) the request waits for the process to complete. By default or when set to 0, the generic request timeout configured in the cluster is applied.
	 */
	requestTimeout?: number
}

export interface ProcessInstanceCreationStartInstruction {
	/**
	 * future extensions might include
	 * - different types of start instructions
	 * - ability to set local variables for different flow scopes
	 * for now, however, the start instruction is implicitly a
	 * "startBeforeElement" instruction
	 */
	elementId: string
}

export interface CreateProcessInstanceFromProcessDefinitionId<
	V extends JSONDoc | LosslessDto,
> extends CreateProcessBaseRequest<V> {
	/**
	 * the BPMN process ID of the process definition
	 */
	processDefinitionId: string
}

export interface CreateProcessInstanceFromProcessDefinition<
	V extends JSONDoc | LosslessDto,
> extends CreateProcessBaseRequest<V> {
	/**
	 * the key of the process definition
	 */
	processDefinitionKey: string
}

export type CreateProcessInstanceReq<T extends JSONDoc | LosslessDto> =
	| CreateProcessInstanceFromProcessDefinitionId<T>
	| CreateProcessInstanceFromProcessDefinition<T>

export interface PatchAuthorizationRequest {
	/** The key of the owner of the authorization. */
	ownerKey: string
	/** Indicates if permissions should be added or removed. */
	action: 'ADD' | 'REMOVE'
	/** The type of resource to add/remove perissions to/from. */
	resourceType:
		| 'AUTHORIZATION'
		| 'MESSAGE'
		| 'JOB'
		| 'APPLICATION'
		| 'TENANT'
		| 'DEPLOYMENT'
		| 'PROCESS_DEFINITION'
		| 'USER_TASK'
		| 'DECISION_REQUIREMENTS_DEFINITION'
		| 'DECISION_DEFINITION'
		| 'USER_GROUP'
		| 'USER'
		| 'ROLE'
	/** The permissions to add/remove. */
	permissions: {
		/** Specifies the type of permissions. */
		permissionType: 'CREATE' | 'READ' | 'UPDATE' | 'DELETE'
		/** A list of resource IDs the permission relates to. */
		resourceIds: []
	}[]
}
