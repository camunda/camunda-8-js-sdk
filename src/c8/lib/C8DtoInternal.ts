import { Int64String, LosslessDto } from '../../lib'

import {
	DecisionDeployment,
	DecisionRequirementsDeployment,
	FormDeployment,
	ProcessDeployment,
} from './C8Dto'

export class DeployResourceResponseDto extends LosslessDto {
	@Int64String
	deploymentKey!: string
	deployments!: (
		| { processDefinition: ProcessDeployment }
		| { decisionDefinition: DecisionDeployment }
		| { decisionRequirements: DecisionRequirementsDeployment }
		| { form: FormDeployment }
	)[]
	tenantId!: string
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Ctor<T> = new (obj: any) => T // eslint-disable-next-line @typescript-eslint/no-explicit-any

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type UnknownRequestBody = Record<string, any> // Interface for requests that return JSON (json=true or undefined)

export interface JsonApiEndpointRequest<
	T extends UnknownRequestBody = UnknownRequestBody,
> extends BaseApiEndpointRequest<T> {
	/** JSON-parse response? Default: true */
	json?: true | undefined
}
// Interface for requests that return raw response (json=false)

export interface RawApiEndpointRequest<
	T extends UnknownRequestBody = UnknownRequestBody,
> extends BaseApiEndpointRequest<T> {
	/** JSON-parse response? */
	json: false
}
// Combined type for use in function signatures

export type ApiEndpointRequest<
	T extends UnknownRequestBody = UnknownRequestBody,
> = JsonApiEndpointRequest<T> | RawApiEndpointRequest<T>

export type SearchSortRequest<T = string> = Array<{
	/** The field to sort by. */
	field: T
	/** The order in which to sort the related field. Default value: ASC */
	order?: 'ASC' | 'DESC'
}> // Base interface with common properties

interface BaseApiEndpointRequest<
	T extends UnknownRequestBody = UnknownRequestBody,
> {
	method: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH'
	/** The URL path of the API endpoint. */
	urlPath: string
	/** The request body. */
	body?: T
	/** TODO: multi-part form support needs to be implemented */
	formData?: FormData
	/** The query parameters. */
	queryParams?: { [key: string]: string | number | boolean | undefined }
	/** The headers. */
	headers?: { [key: string]: string | number | boolean }
	/** A custom JSON parsing function */
	parseJson?: typeof JSON.parse
}
