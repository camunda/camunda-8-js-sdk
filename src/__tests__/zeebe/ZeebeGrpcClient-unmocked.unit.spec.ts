import { EnvironmentSetup, EnvironmentStorage } from '../../lib'
import { allowAny } from '../../test-support/testTags'
import { ZeebeGrpcClient } from '../../zeebe'

let storage: EnvironmentStorage = {}

beforeAll(() => (storage = EnvironmentSetup.storeEnv()))
beforeEach(() => {
	EnvironmentSetup.wipeEnv()
})
afterAll(() => {
	EnvironmentSetup.restoreEnv(storage)
})

vi.setConfig({ testTimeout: 13_000 })

test.runIf(
	allowAny([
		{ deployment: 'unit-test' },
		{ deployment: 'saas' },
		{ deployment: 'self-managed' },
	])
)(
	'ZeebeGrpcClient constructor throws an exception when there is no broker and retry is false',
	async () => {
		const zbc = new ZeebeGrpcClient({
			config: {
				CAMUNDA_OAUTH_DISABLED: true,
				ZEEBE_GRPC_ADDRESS: 'localhoster',
				zeebeGrpcSettings: {
					ZEEBE_GRPC_CLIENT_RETRY: false,
				},
			},
		})
		expect.assertions(1)
		try {
			await zbc.deployResource({
				processFilename: './src/__tests__/testdata/hello-world.bpmn',
			})
		} catch (e: unknown) {
			expect((e as Error).message.indexOf('14 UNAVAILABLE:')).toEqual(0)
		}
		await zbc.close()
	}
)

// We don't really need to test this anymore. It's a known behaviour
test.skip('cancelProcessInstance throws an exception when processInstanceKey is malformed', async () => {
	const zbc = new ZeebeGrpcClient({
		config: {
			CAMUNDA_OAUTH_DISABLED: true,
			ZEEBE_GRPC_ADDRESS: 'localhoster',
			zeebeGrpcSettings: { ZEEBE_GRPC_CLIENT_RETRY: false },
		},
	})
	expect.assertions(1)
	try {
		await zbc.cancelProcessInstance('hello-world')
	} catch (e: unknown) {
		expect(e).toMatchSnapshot()
	}
	await zbc.close()
})
