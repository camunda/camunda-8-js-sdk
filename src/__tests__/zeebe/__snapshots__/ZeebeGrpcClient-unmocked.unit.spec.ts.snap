// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cancelProcessInstance throws an exception when processInstanceKey is malformed 1`] = `
[Error: 
		  processInstanceKey is malformed, value : hello-world
		  ]
`;

exports[`cancelProcessInstance throws an exception when workflowInstanceKey is malformed 1`] = `
[Error: 
		  processInstanceKey is malformed, value : hello-world
		  ]
`;
